hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go

```diff
@@ -2328,6 +2360,311 @@ func (r *HostedClusterReconciler) reconcileAutoscaler(ctx context.Context, creat
 	return autoscaler.ReconcileAutoscaler(ctx, r.Client, hcp, clusterAutoscalerImage, utilitiesImage, createOrUpdate, r.SetDefaultSecurityContext, config.MutatingOwnerRefFromHCP(hcp, releaseVersion))
 }
 
+type etcdJobStatus struct {
+	exists     bool
+	finished   bool
+	successful bool
+}
+
+func (r *HostedClusterReconciler) reconcileETCDMemberRecovery(ctx context.Context, hcluster *hyperv1.HostedCluster, createOrUpdate upsert.CreateOrUpdateFN) (*time.Duration, error) {
+	log := ctrl.LoggerFrom(ctx)
+	hcpNS := manifests.HostedControlPlaneNamespace(hcluster.Namespace, hcluster.Name)
+
+	// Preliminary checks to see if we need to proceed with the recovery procedure
+	if hcluster.Spec.ControllerAvailabilityPolicy != hyperv1.HighlyAvailable {
+		return nil, nil
+	}
+
+	// Check the recovery job
+	recoveryJob := etcdrecoverymanifests.EtcdRecoveryJob(hcpNS)
+	jobStatus, err := r.etcdRecoveryJobStatus(ctx, recoveryJob)
```

Comment:
why is this a job vs a long running deployment which works in loop?

Reply:
The reason this is done from the HO is that we want to be able to recover even older clusters that do not necessarily have the latest CPO (it's a managed service requirement)
There's a few benefits for running a job:
- we're not adding a new permanent workload to control planes which can impose new resource requirements per cp
- a failed job and associated pods makes it easier to 1) understand that recovery was tried, 2) look at the logs of each attempt to understand why it failed. It's not as obvious from a controller to understand what happened unless you look at the controller logs.